### PR TITLE
Brief snippet docs, Wrap comint mode

### PR DESCRIPTION
**Pull request by russmatney**
I can break these apart if you'd rather treat them separately.

Just some thoughts, feel free to take it or leave it. Many thanks for
the work on this mode, it has made learning ink completely within emacs
a great experience. The snippets and auto-completion are a delight!

--

feat: wrap comint with ink-play derived mode 

This wraps the comint-mode buffer used for running ink-play with a
derived mode called ink-play-mode, which adds a single keybinding for
jumping to the docs.

This isn't really necessary, but while reading the manual and testing
out examples, I was running examples and then wanting to jump back to
the docs. It's certainly possible to just set a keybinding elsewhere,
jump via buffer, or use `M-x ink-display-manual`, but I've seen comint
wrapped a few different ways recently and had never done it myself, so
took a shot at it here.

Ink-play mode could be expanded to rewind choices or restart the last
knot/stitch played at some point. Another idea could be writing a
binding to jump to the current choice/prompt from the 'play' buffer.

-- 

docs: add snippets section with keys

I was going through these myself to learn them, so thought I'd throw in
a docs section - feel free to take/leave/edit as you would.

### CODE REVIEW

**Comment by PiOverFour** [comment]
Hi @russmatney, sorry for the long answering time. Thanks for the interest, I’m glad you found the mode useful! :)
I think wrapping the comint is a good idea, it’s one of the areas I haven’t fully grasped when working on ink-mode.
There are a few things I’m not getting upon testing, though.
- In my setup, two identical submenus appear:
image
- What is line 785 about? It appears as is in the mode documentation, so it doesn’t seem to be intentional.
``` elisp
//{ink-play-mode-map}")
```
> Ink-play mode could be expanded to rewind choices or restart the last knot/stitch played at some point.

I’m not sure how that would work, since it would require to save states in inklecate, and restore it.

> Another idea could be writing a binding to jump to the current choice/prompt from the 'play' buffer.

Similarly, we’d need to know from inklecate which point we’re at in the story. Do you know how this could be achieved?

**Comment by russmatney** [comment]
No worries on the timing! Thanks much for the time you've put into this project
already :)

Also, thanks for taking the time to test my branch yourself - I'm glad we caught
some of these ahead of time. I was extending myself to get this working, and
could have cleaned up/learned more before opening this, but was pretty happy when
it worked so just pushed it up and opened this rather hastily.

The documentation syntax you called out definitely looks odd - I saw it in a few
places while digging around comint mode and others, and copied it in at one
point, thinking some voodoo was required to get keybindings in place - later I
realized it was the defvar not updating my value for ink-play-mode-map, and
neglected to remove this. Some digging reveals it provides some helpers to
describe-mode - see 'Auto Generated Keybinding List'
[here](http://ergoemacs.org/emacs/inline_doc.html). I'm open to leaving or
dropping it, I'm not sure how relevant that convention is anymore.

The doubling of the ink-play mode is strange - I'm not sure what might be
causing that. I'm not familiar with easy-menu, maybe that's related? I only
included it to match the ink-mode creation code. I usually use emacs with no top
menu - I can try some other ways to reproduce it.

I added another commit that expands the ink-play function to work from
ink-play-mode itself, allowing for a quick restart with the last file and/or
last knot played, using a var as a cache for each. There's more detail in that
commit message. Let me know what you think!

The features I mentioned were not well thought out - I was more reaching for a
defense of wrapping comint mode, and these seemed like things that would be nice
to have. Mostly, there's a degree to which I felt like I was getting stuck in
the comint buffer, and having to navigate away to get the ink-mode power back.

I'm new to ink and not too familiar with inklecate, but I'm planning on
integrating ink into a game in the next week, so will dig in further. I see what
you're saying about needing a state from inklecate for this to be simpler. One
thought toward a rewind feature was that we could record the inputs from the
user, pop decisions off of that, then replay all but the last input to simulate
a rewind... but that feels like a headache to manage, really, and I'm not sure
it's really necessary after the replay feature.

Similarly, jumping from the play buffer at a specific choice back to a
particular location in the script could be done by parsing the inklecate output
and finding a matching line in the script... but the match wouldn't be perfect,
so this might end up being more annoying that it's worth. Currently we can just
navigate back to the file and search it by hand. Though, maybe there's value in
a binding that jumps to whatever file is in `ink-play-last-file-name`, or
straight to the last called knot, if there is one.

One other note: I matched the ink-play and ink-play-knot keybindings in
ink-play-mode, which means `C-c C-c` is now more of a restart than comint's
usual 'interrupt'. I don't mind this behavior, but thought it should be called
out.

**Comment by PiOverFour** [comment]
I had some thinking, and I’m not really convinced that showing a list of snippets in the README is very useful, since they are kind of self-documenting: you can get a list from `yas-insert-snippet` and from the YASnippet menu if you use that…  
A more concise way of advertising them in the README would be to create a link to the directory itself. I tested that in [a branch](https://github.com/PiOverFour/ink-mode/tree/test_branch_link#features) and it is supported by GitHub.  
What do you think?

> The documentation syntax you called out definitely looks odd […]
> Some digging reveals it provides some helpers to
> describe-mode - see 'Auto Generated Keybinding List'
> [here](http://ergoemacs.org/emacs/inline_doc.html). I'm open to leaving or
> dropping it, I'm not sure how relevant that convention is anymore.

It doesn’t seem to work as far as I can tell: the key bindings are already displayed anyway, and the hint itself is printed. I’d rather leave it out.
image

> The doubling of the ink-play mode is strange - I'm not sure what might be
> causing that. I'm not familiar with easy-menu, maybe that's related? I only
> included it to match the ink-mode creation code. I usually use emacs with no top
> menu - I can try some other ways to reproduce it.

I don’t use the menu either and included it to be comprehensive while learning about mode creation :D  
It’s bizarre indeed. I tried to investigate it a bit, but found nothing wrong with the way it is defined.  
If we can’t fix it, we should leave it out for now.

> I added another commit that expands the ink-play function to work from
> ink-play-mode itself, allowing for a quick restart with the last file and/or
> last knot played, using a var as a cache for each. There's more detail in that
> commit message. Let me know what you think!

Nice idea :)

> Mostly, there's a degree to which I felt like I was getting stuck in
> the comint buffer, and having to navigate away to get the ink-mode power back.

I see. I think I had the same problem: I wanted to make the buffer exitable when the Ink process has finished. Some modes (splash screen, info, etc.) have a convenient `q` keybinding to close their buffers, and that would be nice to have for `ink-play`. But I could never find out how to implement it in the comint buffer, but activate the keybinding only when the Ink process has finished. Maybe now it will be easier!

> One thought toward a rewind feature was that we could record the inputs from the
> user, pop decisions off of that, then replay all but the last input to simulate
> a rewind...

I don’t think it’s really feasible either, because you may have random dialogues, events, or even branching in Ink, so you wouldn’t necessarily be able to consistently get the same play-through twice with the same input. Anyway, it’s probably overkill anyway.

> One other note: I matched the ink-play and ink-play-knot keybindings in
> ink-play-mode, which means `C-c C-c` is now more of a restart than comint's
> usual 'interrupt'. I don't mind this behavior, but thought it should be called
> out.

I think it’s a good choice as well. I usually kill the Ink buffer when I need to interrupt it anyway!

----

### TL;DR
* [ ] I’d rather not have the full snippet list in the README;
* [ ] Remove doc syntax;
* [ ] Fix duplicate menu, or remove it if it proves impossible;
* [ ] A few inline style comments.

File: ink-mode.el [hunk 1]
```diff
@@ -780,24 +806,66 @@ indent. INDENTATION is the current sum."
...
+                      (buffer-file-name)))
+
+         ;; needs to run before set-buffer call
+         (knot-name (when go-to-knot
+                      (if in-ink-play-buffer
+                           ;; return the last played knot
+                           ink-last-played-knot
+                           ;; return a knot name if found, nil otherwise
+                           (let ((x (ink-get-knot-name)))
+                             (unless (eq "" x) x)))))
```
Comment (PiOverFour) [inline ink-mode.el:835]: Please reindent this.

**Comment by russmatney** [comment]
Thanks for taking the time to review!

I've updated each of the things you called out, let me know if there's any more.

### re: snippet dir-link vs table

`yas-insert-snippet` is quite useful to call directly, thanks for that. I wish it included the shorthand for each snippet as well, but meh. The link in the docs is fine, no strong opinion either way. I've reverted the related readme commit on this branch for now. I can rework/rebase the commits to remove that touch completely if you prefer.

### re: q to kill the ink-play-mode buffer (when the process is finished).

Seems reasonable. An approach I'd try would be to use `(comint-check-proc "*Ink*")` similar to the ink-play fn - no-op if there is a running process, quit otherwise.

--

I'm making an effort to write interactive fiction for nanowrimo this month, all in ink-mode! The usage is sparking more ideas, so I may have more stuff coming - in particular I miss some of the bindings I'm used to from org-mode, such as ctrl-enter to create a new-bullet-of-same-type-below (`+org/insert-item-below`). I'm also thinking about a command to cycle between bullet types... These are things I haven't implemented in emacs before, so no promises yet :)

**Comment by PiOverFour** [comment]
> The link in the docs is fine, no strong opinion either way. I've reverted the related readme commit on this branch for now. I can rework/rebase the commits to remove that touch completely if you prefer.

I rewrote the history for your branch, squashing reverts together. For now it’s at [my repo](https://github.com/PiOverFour/ink-mode/commits/russ-rebase), I’ll merge it if that’s all right with you.

> I'm making an effort to write interactive fiction for nanowrimo this month, all in ink-mode!

That’s cool, I didn’t know about this event!

> I may have more stuff coming - in particular I miss some of the bindings I'm used to from org-mode, such as ctrl-enter to create a new-bullet-of-same-type-below (`+org/insert-item-below`). I'm also thinking about a command to cycle between bullet types... These are things I haven't implemented in emacs before, so no promises yet :)

These are good ideas, feel free to implement them—or not, no pressure!

Thanks for your work on this! :)

**Comment by russmatney** [comment]
Sounds good, feel free to merge from your repo.

**Comment by PiOverFour** [comment]
Cool, this was just merged!
